File: backend/src/utils/activityLogger.js
const { getPool, sql } = require('../config/database');

// Không throw lỗi ra ngoài — ghi log thất bại không được làm hỏng thao tác chính
const logActivity = async ({
  maNguoiDung = null, tenDangNhap = '', hanhDong, doiTuong = null, maDoiTuong = null,
  moTa = '', ketQua = 'THANH_CONG', lyDoThatBai = null,
}) => {
  try {
    const pool = getPool();
    await pool.request()
      .input('maNguoiDung', sql.Int, maNguoiDung)
      .input('tenDangNhap', sql.VarChar(50), tenDangNhap)
      .input('hanhDong', sql.VarChar(50), hanhDong)
      .input('doiTuong', sql.VarChar(100), doiTuong)
      .input('maDoiTuong', sql.VarChar(50), maDoiTuong)
      .input('moTa', sql.NVarChar, moTa)
      .input('ketQua', sql.VarChar(20), ketQua)
      .input('lyDoThatBai', sql.NVarChar, lyDoThatBai)
      .query(`INSERT INTO NhatKyHoatDong (maNguoiDung, tenDangNhap, hanhDong, doiTuong, maDoiTuong, moTa, ketQua, lyDoThatBai)
              VALUES (@maNguoiDung, @tenDangNhap, @hanhDong, @doiTuong, @maDoiTuong, @moTa, @ketQua, @lyDoThatBai)`);
  } catch (err) {
    console.error('Lỗi ghi nhật ký hoạt động:', err.message);
  }
};

module.exports = { logActivity };

File: backend-dotnet/Services/ActivityLogger.cs
using backend_dotnet.Models;

namespace backend_dotnet.Services;

public interface IActivityLogger
{
    Task LogAsync(
        int? maNguoiDung, string? tenDangNhap, string hanhDong,
        string? doiTuong = null, string? maDoiTuong = null, string moTa = "",
        string ketQua = "THANH_CONG", string? lyDoThatBai = null);
}

// Tương đương backend/src/utils/activityLogger.js — ghi log thất bại không được làm hỏng thao tác chính.
public class ActivityLogger(QuanLyKhoQuanKhiContext db, ILogger<ActivityLogger> logger) : IActivityLogger
{
    public async Task LogAsync(
        int? maNguoiDung, string? tenDangNhap, string hanhDong,
        string? doiTuong = null, string? maDoiTuong = null, string moTa = "",
        string ketQua = "THANH_CONG", string? lyDoThatBai = null)
    {
        try
        {
            db.NhatKyHoatDongs.Add(new NhatKyHoatDong
            {
                MaNguoiDung = maNguoiDung,
                TenDangNhap = tenDangNhap,
                HanhDong = hanhDong,
                DoiTuong = doiTuong,
                MaDoiTuong = maDoiTuong,
                MoTa = moTa,
                ThoiGian = DateTime.Now,
                KetQua = ketQua,
                LyDoThatBai = lyDoThatBai,
            });
            await db.SaveChangesAsync();
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Loi ghi nhat ky hoat dong");
        }
    }
}
